endpoint: Do not skip datapath rewrites on duplicate regenerations

jrajahalme opened the pull request:
Store the highest skipped regeneration level when skipping a duplicate endpoint regeneration, so that the queued regeneration can be performed on the required level.

```release-note
Do not skip datapath rewrites when an otherwise duplicate endpoint regeneration requires it.
```

jrajahalme committed:
endpoint: Remove pointless setState()

Replace setState() with the current state with the corresponding
status log to make it clearer that the state is not being changed in
this case.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
endpoint: Default regeneration reason to "Invalid"

Force explicit initialization of regeneration reason to avoid
defaulting to regeneration without datapath.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -833,7 +833,8 @@ func (e *Endpoint) Update(cfg *models.EndpointConfigurationSpec) error {
 	regenCtx := &regeneration.ExternalRegenerationMetadata{
-		Reason: "endpoint was updated via API",
+		Reason:            "endpoint was updated via API",
+		RegenerationLevel: regeneration.RegenerateWithoutDatapath, // Needs datapath load??
```
The code immediately below this statement already handles the cases where we're worried about a higher regeneration level, so this can probably regenerate without the datapath. (Assuming the `updateAndOverrideEndponitOptions()` does what it sounds like).

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -1025,7 +1026,10 @@ func (e *Endpoint) leaveLocked(proxyWaitGroup *completion.WaitGroup, conf Delete
-	if !<-e.Regenerate(&regeneration.ExternalRegenerationMetadata{Reason: reason}) {
+	if !<-e.Regenerate(&regeneration.ExternalRegenerationMetadata{
+		Reason:            reason,
+		RegenerationLevel: regeneration.RegenerateWithoutDatapath, // Needs datapath load??
```
This seems to be only used from the following API:

`PATCH /endpoint/{id} request`

Who knows what someone could shove in there, probably best to 
```suggestion
		RegenerationLevel: regeneration.RegenerateWithDatapathRewrite,
```

jrajahalme replied:
Refactored with this change.

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -1794,7 +1798,10 @@ func (e *Endpoint) identityLabelsChanged(ctx context.Context, myChangeRev int) e
-		e.Regenerate(&regeneration.ExternalRegenerationMetadata{Reason: "updated security labels"})
+		e.Regenerate(&regeneration.ExternalRegenerationMetadata{
+			Reason:            "updated security labels",
+			RegenerationLevel: regeneration.RegenerateWithoutDatapath, // Needs datapath load??
```
Identity is a static variable substituted in the ELF:

https://github.com/cilium/cilium/blob/2845ccdfcdbb53a80998a9ea06ddb5f8ff1134ca/pkg/datapath/loader/template.go#L217

So we need to do an ELF rewrite here.

```suggestion
			RegenerationLevel: regeneration.RegenerateWithDatapathRewrite,
```

jrajahalme replied:
Refactored with this change.

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -2143,8 +2150,9 @@ func (e *Endpoint) RegenerateAfterCreation(ctx context.Context, syncBuild bool)
-			ParentContext: ctx,
+			Reason:            "Initial build on endpoint creation",
+			ParentContext:     ctx,
+			RegenerationLevel: regeneration.RegenerateWithoutDatapath, // Needs datapath load??
```
When an endpoint is created it needs to be regenerated at least from the template. Rewrite will take care of this.

```suggestion
			RegenerationLevel: regeneration.RegenerateWithDatapathRewrite,
```

jrajahalme replied:
Refactored with this change.

joestringer commented on pkg/endpoint/regenerationcontext.go:
```diff
@@ -46,6 +47,10 @@ type regenerationContext struct {
 
 func ParseExternalRegenerationMetadata(ctx context.Context, c context.CancelFunc, e *regeneration.ExternalRegenerationMetadata) *regenerationContext {
+	if e.RegenerationLevel == regeneration.Invalid {
+		log.WithField(logfields.Reason, e.Reason).Errorf("Uninitialized regeneration level")
```
Please add this to the CI log failures as well so we don't ignore it.

jrajahalme replied:
Done!

joestringer commented on pkg/endpoint/restore.go:
```diff
@@ -166,7 +166,8 @@ func (e *Endpoint) RegenerateAfterRestore() error {
 	regenerationMetadata := &regeneration.ExternalRegenerationMetadata{
-		Reason: "syncing state to host",
+		Reason:            "syncing state to host",
+		RegenerationLevel: regeneration.RegenerateWithoutDatapath, // Needs datapath load??
```
During restore we don't know what the state of the datapath is so we at least need to rewrite.

```suggestion
		RegenerationLevel: regeneration.RegenerateWithDatapathRewrite,
```

jrajahalme replied:
Changed as indicated.

joestringer commented on pkg/endpoint/restore_test.go:
```diff
@@ -90,7 +90,8 @@ func (ds *EndpointSuite) endpointCreator(id uint16, secID identity.NumericIdenti
 	regenerationMetadata = &regeneration.ExternalRegenerationMetadata{
-		Reason: "test",
+		Reason:            "test",
+		RegenerationLevel: regeneration.RegenerateWithoutDatapath, // Needs datapath load??
```
Is this variable used at all?

jrajahalme replied:
Maybe not, but I'd like to keep `RegenerationLevel` explicitly initialized also in tests, in case someone ever copy&pastes it somewhere else.

jrajahalme replied:
Removed the comment.

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -442,6 +450,11 @@ func (e *Endpoint) RegenerateIfAlive(regenMetadata *regeneration.ExternalRegener
+			// Bump the skipped regeneration level if needed so that the existing/queued
+			// regeneration can regenerate on the required level.
+			if regenMetadata.RegenerationLevel > e.skippedRegenerationLevel {
+				e.skippedRegenerationLevel = regenMetadata.RegenerationLevel
```
Maybe we should do a similar `logStatusLocked` here but log the priority bump of the regeneration level? It may be useful to know if this logic triggers or not.

Probably `if/else` with the line immediately below.

jrajahalme replied:
Done!

joestringer reviewed (changes requested):
The third patch seems to only change the `RegenerateIfAlive()` case. Did you trace the other callers of `Regenerate` to check whether we need to fold the higher-priority regeneration level in from those paths as well?

joestringer reviewed (changes requested):
Please also either rename the PR title or add a release-note section to add a one-sentence implication of the bug fixed in user-friendly language, the current title sounds innocuous and users will not know that this is the bug causing the issue in their cluster.

jrajahalme committed:
endpoint: Store the highest skipped regeneration level

Store the highest skipped regeneration level so that the regeneration
can be performed on the required level.

This is achieved by refactoring Endpoint.RegenerateIfAlive() into three pieces:

- setRegenerateStateLocked(): Change Endpoint state for regeneration,
  if not already done. If the current state indicates that a
  regeneration is already pending, store the current regenetion level
  to the new Endpoint.skippedRegenerationLevel, if higher so that the
  pending regeneration can be performed at that level.

- SetRegenerateStateIfAlive(): Call setRegenerateStateLocked() if
  endpoint is still alive

- RegenerateIfAlive(): Call SetRegenerateStateIfAlive() and
  Regenerate() if possible.

All other sites that were previously manipulating Endpoint.state for
regeneration are refactored to use one of the above three functions as
appropriate. This allows the regeneration level recording of a skipped
regeneration to be managed in a single function
(setRegenerateStateLocked()) instead of copying the logic all over the
place.

Endpoint.RegenerateAfterCreation() used to condition regeneration on
'e.getState() == StateReady' to avoid regeneting again if the endpoint
has already been regenerated due to Endpoint's labels being received
from the kv-store. However, Daemon.createEndpoint() expects endpoint
regeneration only happen when it finally calls
Endpoint.RegenerateAfterCreation(), after the call to UpdateLabels().
Fix this by refactoring Daemon.createEndpoint() so that endpoint
regeneration is OK right after calling Endpoint.UpdateLabels() and
skipping endpoint regeneration trigger later if it was already
triggered, and possibly already completely performed, thus avoiding
unnecessary duplicate regeneration. Make this more explicit by
inlining Endpoint.RegenerateAfterCreation() into
Daemon.createEndpoint() which was it's only caller anyway.

Suggested-by: Dan Wendlandt <dan@covalent.io>
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
